### ISSUE TITLE
Respect relation items order

### ISSUE THREAD

**Issue by bjnsn** [strapi/strapi#2166]
NOTE: This a duplicate of this (closed) issue that was not submitted with a proper template: 
https://github.com/strapi/strapi/issues/2134

**Informations**
- Node.js version: 10.12.0
- npm version: 6.4.1 
- Strapi version:  3.0.0-alpha.14.3
- Database: MongoDB (4.0.0)
- Operating system: Mac OS 10.13.2

(NOTE: also seeing same behavior on server configured as follows)
- Node.js version: 9.9.0
- npm version: 5.6.0 
- Strapi version:  3.0.0-alpha.14.2
- Database: MongoDB (4.0.0)
- Operating system: Ubuntu 16.04.4

**What is the current behavior?**
Relations do not retain ordering on save.

**Steps to reproduce the problem**
- Create a new strapi
```
strapi new testProject3
# (accept defaults)
cd testProject3
strapi start
```
- Open admin
- Create two Content Types (foo & bar), each with a name (string) field.
- Create a oneToMany relation from bar to foo.
- Create three foos with name values 'foo1', 'foo2', and 'foo3'
- Create a bar with name 'bar1'
- Add each foo to the oneToMany relation in bar1; order them 'foo3', 'foo2', 'foo1'
- Save bar1
- Reopen bar1

The oneToMany relation UI now shows items in order 'foo1', 'foo2', 'foo3'.
NOTE: this is also true with many-to-many relations, as shown in these screenshots.

Before save:
<img width="283" alt="screen shot 2018-10-18 at 10 53 16 am" src="https://user-images.githubusercontent.com/791610/47174947-e741b200-d2c6-11e8-82b6-d1f81e0779ed.png">

After save:
<img width="282" alt="screen shot 2018-10-18 at 10 53 47 am" src="https://user-images.githubusercontent.com/791610/47174992-03455380-d2c7-11e8-82a8-72bb85ac4c2c.png">

**What is the expected behavior?**
The oneToMany items retain their ordering.

**Suggested solutions**
None at this point.

Thanks!

**Comment by pbov** [comment]
Same here. Some more details:

- the relations are saved in the order you added them in the first save. So 1) adding foo3, foo2, foo1 2) save -> works as expected, but rearranging relations won't work after saving.
- in MySQL the order isn't saved in a separate field but in the admin panel it's displayed in the order of the internal IDs (and saved in this way). So it will work for the first save - but rearranging and adding/removing further relations will be very tricky without having a separate order-column.

And: the relation order in REST or graphql is absolut random. Maybe this should get its own ticket?

**Comment by derrickmehaffy** [comment]
I can also confirm the thoughts @pbov is stating, right now Strapi is ordering based on IDs so with MySQL this is likely to be the order in which they are created, since MySQL will return them in the order of the IDs

With Mongo this is basically impossible as the IDs are actually randomly generated UUID strings.

As it was stated GraphQL won't keep their order at least with the current way data is handled.

**Comment by mamartins** [comment]
In Mongo what  I found is that the relation that  has the `"dominant": true`, is the one that saves the array and I can order that side of the relation

**Comment by friendofdog** [comment]
I am experiencing the same issue. The `Testimonial` items below related to a `Customer`, where `Customer` is dominant. Here's the original order:

![screen shot 2019-01-24 at 15 55 36](https://user-images.githubusercontent.com/8337121/51660265-8b3aa800-1ff0-11e9-8c53-64953a4647d9.png)

... and here's the subsequent re-ordering:

![screen shot 2019-01-24 at 15 55 23](https://user-images.githubusercontent.com/8337121/51660365-d94fab80-1ff0-11e9-8713-58b20132cc03.png)

UI recognises the reordering. However, GraphQL does not seem to pick up on this. A call using GraphQL will return the following no matter how `Testimonial` items are reordered:

```
{
  "data": {
    "customer": {
      "title": "some restaurant",
      "testimonials": [
        {
          "title": "entered first",
          "content": "reordered to second"
        },
        {
          "title": "entered second",
          "content": "reorederd to third"
        },
        {
          "title": "entered third",
          "content": "reordered to first"
        }
      ]
    }
  }
}
```

On the other hand, REST seems to work. After the `Testimonial` items are re-ordered, the following response is returned:

```
{
    "title": "some restaurant",
    "testimonials": [
        {
            "title": "entered third",
            "content": "reordered to first"
            ...
        },
        {
            "title": "entered first",
            "content": "reordered to second"
            ...
        },
        {
            "title": "entered second",
            "content": "reorederd to third"
            ...
        }
    ],
    ...
}
```

@mamartins, you are correct: order in the UI is retained only on the `"dominant": true` side. However, as @derrickmehaffy points out, GraphQL does not pick up on this. Calls made in REST work.

**Comment by lauriejim** [comment]
Hello, I lock this issue cause to issue is clearly explain.

**Comment by lauriejim** [comment]
The best way to manage this is to manage the order in your request - especially if you use GraphQL. You will be able to fetch relations data and define an order for these.

**Comment by jorisw** [comment]
Just wanted to stress that this is strictly a Strapi GraphQL problem, as the REST endpoint behaves as desired, at least for the dominant content type.

The reason it works in REST is most likely because the order of related items _does_ exist in data: the natural order of entries in the database table/collection. For example, in SQLite, the order of entries in the `typeA__typeB` table represents the order in which the user added the related items, which REST respects.

The GraphQL layer however appears to insist on a default sort, which irrecoverably overrides the 'natural' entry order coming from the database. My hope would be that this is fixable without waiting on a big general database rewrite in Strapi.

**Comment by lauriejim** [comment]
Thank you for this insight.

**Comment by derrickmehaffy** [comment]
Will be relocking the issue, @jorisw poked me on Slack to see if he could add some info. If anyone else needs the same, please let me know.

**Comment by lauriejim** [comment]
This issue has been mentioned on **Strapi Community**. There might be relevant details there:

https://forum.strapi.io/t/save-relation-items-order-issue-2166/332/1

**Comment by derrickmehaffy** [comment]
Hey all, we are currently moving many feature requests to our new feedback and feature request website to help clean up the GitHub issues and make it easier for us to review bug reports and fix them.

I have moved this feature to the following URL: https://feedback.strapi.io/customization/p/respect-relation-items-order
If you are interested in this feature please feel free to go there and add more comments and/or upvote it there.

For now I will close this and lock it so that all new information goes into our new feedback website.
Thanks!
